interfaces/builtin/cups_control.go

```diff
@@ -19,30 +19,159 @@
+// is connected or not for the peer client process and the print service will
+// mediate admin functionality (ie, the rules in these interfaces allow
+// connecting to the print service, but do not implement enforcement rules; it
+// is up to the print service to provide enforcement).
 const cupsControlSummary = `allows access to the CUPS control socket`
 
+// cups-control is implicit on classic but may also be provided by an app snap
+// on core or classic (the current design allows the snap provider to slots
+// both cups-control and cups or just cups-control (like with implicit classic
+// or any slot provider without mediation patches), but not just cups).
 const cupsControlBaseDeclarationSlots = `
   cups-control:
     allow-installation:
       slot-snap-type:
+        - app
         - core
     deny-auto-connection: true
+    deny-connection:
+      on-classic: false
+`
+
+const cupsControlPermanentSlotAppArmor = `
+# Allow daemon access to create the CUPS socket
+/{,var/}run/cups/ rw,
+/{,var/}run/cups/** rwk,
+
+# Allow cups to verify passwords directly
+/etc/shadow r,
+/var/lib/extrausers/shadow r,
+
+# Some versions of CUPS will verify the connecting pid's
+# security label
+@{PROC}/[0-9]*/attr/current r,
+
+# Allow daemon access to the color manager on the system
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/ColorManager
+    interface=org.freedesktop.ColorManager
+    peer=(name=org.freedesktop.ColorManager),
+
+# Allow daemon to send notifications
+dbus (send)
+    bus=system
+    path=/org/cups/cupsd/Notifier
+    interface=org.cups.cupsd.Notifier
+    peer=(name=org.freedesktop.DBus,label=unconfined),
+
+# Allow daemon to send signals to its snap_daemon processes
+capability kill,
+
+# Allow daemon to manage snap_daemon files and directories
+capability fsetid,
+`
+
+const cupsControlConnectedSlotAppArmor = `
+# Allow daemon to send notifications to connected snaps
+dbus (send)
+    bus=system
+    path=/org/cups/cupsd/Notifier
+    interface=org.cups.cupsd.Notifier
+    peer=(name=org.freedesktop.DBus,label=###PLUG_SECURITY_TAGS###),
 `
 
 const cupsControlConnectedPlugAppArmor = `
-# Description: Can access cups control socket. This is restricted because it provides
-# privileged access to configure printing.
+# Allow communicating with the cups server for printing and configuration.
 
 #include <abstractions/cups-client>
-/run/cups/printcap r,
+/{,var/}run/cups/printcap r,
+
+# Allow receiving all DBus signal notifications from the daemon (see
+# notifier/dbus.c in cups sources)
+dbus (receive)
+    bus=system
+    path=/org/cups/cupsd/Notifier
+    interface=org.cups.cupsd.Notifier
+    peer=(name=org.freedesktop.DBus,label=###SLOT_SECURITY_TAGS###),
 `
 
+type cupsControlInterface struct {
+	commonInterface
+}
+
+func (iface *cupsControlInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	// Only apply slot snippet when running as application snap on classic
```

Comment:
this comment is confusing,  on core the helper is always false, which means we apply the snippet. I think the comment would be clearer as:

// On classic only apply slot snippet when...